Rust Secure Coding Guidelines (#137)
aptos-labs/developer-docs: docs/community/rust-coding-guidelines/rust-secure-coding-guidelines.md

Clippy with Aptos-specific configuration can be run locally via `cargo xclippy` or using rust-analyser in your preferred IDE following these [instructions](https://rust-analyzer.github.io/manual.html#clippy). Aptos uses directives in files and a per-directory configuration to turn on or off checks.

### Rustfix

Apply `rustfix` for compiler warnings and edition transitions, but verify the automatic fixes to ensure that the recommendations match the purpose of the code.

### Documentation

Document safety invariants and security considerations in code, especially for public and `unsafe` functions.

## Libraries

### Crate Quality and Security

Assess and monitor the quality and maintenance of crates that are being introduced to the codebase, employing tools like `cargo-outdated` and `cargo-audit` for version management and vulnerability checking.

- Aptos utilizes **[Dependabot](https://github.com/dependabot)** to continuously monitor libraries. Our policy requires mandatory updates for critical and high-vulnerabilities, or upon impact evaluation given the context for medium and lower.

> - We recommend leveraging [deps.dev](https://deps.dev) to evaluate new third party crates. This site provides an OpenSSF stcorecard containing essential information. As a guideline, libraries with a score of 7 or higher are typically safe to import. However, those scoring **below 7** must be flagged during the PR and require a specific justification.

Reviewer: typo `stcorecard`

---

> id: rust-secure-coding-guidelines

Reviewer 2: should this be `aptos-secure-coding-guidelines` ?

> title: Secure Coding

Reviewer 2: should this be Secure Coding for Aptos Core or something along those lines?

---

These Rust Secure Coding Guidelines are essential for anyone contributing to Aptos, reflecting our security-first approach. As Aptos is built with a primary focus on security, these guidelines, derived and adapted from ANSSI's Secure Rust Guidelines, are integral to maintaining the high standards of safety and robustness in aptos-core. Aptos contributors are encouraged to thoroughly understand and apply these principles in their work.

## Development Environment

### Rustup

> Utilize Rustup for managing Rust toolchains. However, keep in mind that, from a security perspective, Rustup performs all downloads over HTTPS, but it does not yet validate signatures of downloads. Security is shifted to [create.io](http://create.io) and GitHub repository hosting the code [[rustup]](https://anssi-fr.github.io/rust-guide/02_devenv.html#rustup).

Reviewer 2: I think we should remove per-item references to the ANSSI guidelines and only link them at the end as references. We say at the top that our guidelines are based on the ANSSI ones, we don't have to match them.

### Stable Toolchain

> Development of a secure application must be done using a fully stable toolchain, for limiting potential compiler, runtime, or tool bugs [[toolchains considerations]](https://anssi-fr.github.io/rust-guide/02_devenv.html#stable-nightly-and-beta-toolchains).

Reviewer 2: ```suggestion
Aptos Core leverages Rust stable toolchain to limit potential compiler, runtime, or tooling bugs, or potential supply chain attacks in nightly releases.
```

Clippy with Aptos-specific configuration can be run locally via `cargo xclippy` or using rust-analyser in your preferred IDE following these [instructions](https://rust-analyzer.github.io/manual.html#clippy). Aptos uses directives in files and a per-directory configuration to turn on or off checks.

### Rustfix

Apply `rustfix` for compiler warnings and edition transitions, but verify the automatic fixes to ensure that the recommendations match the purpose of the code [[rustfix]](https://anssi-fr.github.io/rust-guide/02_devenv.html#rustfix).

### Documentation

Document safety invariants and security considerations in code, especially for public and `unsafe` functions.

````rust
  foo(
      // SAFETY:
      // This is a valid safety comment
      unsafe { *x }
  )
  ```

  ```rust
  use std::ptr::NonNull;
  let a = &mut 42;

  // SAFETY: references are guaranteed to be non-null.
  let ptr = unsafe { NonNull::new_unchecked(a) };
````

## Libraries

### Crate Quality and Security

> Assess and monitor the quality and maintenance of used crates, employing tools like `cargo-outdated` and `cargo-audit` for version management and vulnerability checking [[audit lib]](https://anssi-fr.github.io/rust-guide/03_libraries.html#libraries).

Reviewer 2: ```suggestion
Assess and monitor the quality and maintenance of crates that are being introduced to the codebase, employing tools like `cargo-outdated` and `cargo-audit` for version management and vulnerability checking [[audit lib]](https://anssi-fr.github.io/rust-guide/03_libraries.html#libraries).
```

- Aptos utilizes **[Dependabot](https://github.com/dependabot)** to continuously monitor libraries. Our policy requires mandatory updates for critical and high-vulnerabilities, or upon impact evaluation given the context for medium and lower.

> - When vetting a new library, it's advisable to utilize deps.dev for evaluation. This site provides an OpenSSF scorecard containing essential information. As a guideline, libraries with an OpenSSF score of 7 or higher are typically safe to import. However, those scoring **below 7** must undergo discussion prior to importation.

Reviewer 2: ```suggestion
- We recommend leveraging [deps.dev](https://deps.dev) to evaluate new third party crates. This site provides an OpenSSF stcorecard containing essential information. As a guideline, libraries with a score of 7 or higher are typically safe to import. However, those scoring **below 7** must be flagged during the PR and require a specific justification.
```

### Minimize Use of Feature Flags

As a general practice, avoid using feature flags in your crates unless absolutely necessary. Feature flags can introduce complexity and unexpected behaviours, making the codebase harder to audit for security vulnerabilities.

### Understanding Feature Unification

Be aware of Cargo's feature unification process. When multiple dependencies require the same crate with different feature flags, Cargo unifies these into a single configuration. This unification can inadvertently enable features that might not be desirable or secure for the project [[Rustbook: features unification]](https://doc.rust-lang.org/cargo/reference/features.html#feature-unification) [[Rustbook: feature resolver]](https://doc.rust-lang.org/cargo/reference/features.html#feature-resolver-version-2).

## Language Generalities

### Unsafe Code

> Never use `unsafe` blocks unless as a last resort. Justify their use in a comment, detailing how the code is effectively safe to deploy [[unsafe code]](https://anssi-fr.github.io/rust-guide/04_language.html#unsafe-code).

Reviewer 2: should the "Documentation" section about documenting safe code be moved here?

Author: "Document safety invariants and security considerations in code, especially for public and `unsafe` functions." it's covers also invariants, we can reiterate it in this section.

### Integer Overflows

> Refer to [coding-guidelines](./rust-coding-guidelines.md#integer-arithmetic).

Reviewer 2: this moves the attention to another document. can we replicate/reframe the content here with a security focus?

Author: The idea was to don't duplicate material between coding guidelines and secure coding.

### Error Handling

Use `Result<T, E>` and `Option<T>` for error handling instead of _unwrapping_ or _expecting_, to avoid panics [[error handling]](https://anssi-fr.github.io/rust-guide/04_language.html#error-handling) [[panic]](https://anssi-fr.github.io/rust-guide/04_language.html#panics) [[coding-style]](rust-coding-guidelines.md/#error-handling).

### Assertions

Prefer using `Result` and context-rich error handling over Rust's `assert!`, `assert_eq!`, and `assert_ne!` macros for enforcing invariants, reserving assertions for development and unrecoverable error scenarios.

## Types Systems and Data Structures

### Drop Trait

> Implement the `Drop` trait selectively and ensure it doesn't cause panics [[drop trait the destructor]](https://anssi-fr.github.io/rust-guide/06_typesystem.html#drop-trait-the-destructor).

Reviewer 2: in which circumstances it can cause panics?

> Implement the `Drop` trait selectively and ensure it doesn't cause panics [[drop trait the destructor]](https://anssi-fr.github.io/rust-guide/06_typesystem.html#drop-trait-the-destructor).

Reviewer 2: ```suggestion```

> Implement the `Drop` trait selectively and ensure it doesn't cause panics [[drop trait the destructor]](https://anssi-fr.github.io/rust-guide/06_typesystem.html#drop-trait-the-destructor).

Reviewer 2: this sentence effectively repeats what's also below, so I would remove it.

- Implement the Drop trait selectively, only when necessary for specific destructor logic. It's mainly used for managing external resources or memory in structures like Box or Rc, often involving unsafe code and security-critical operations.
- In a Rust secure development, the implementation of the `std::ops::Drop` trait
  must not panic.
- Do not rely on `Drop` trait in security material treatment after the use, use [zeroize](https://docs.rs/zeroize/latest/zeroize/#) to explicit destroy security material, e.g. private keys.

> ### Send and Sync Traits

Reviewer 2: how much is this relevant for the existing codebase?

Author: We have some instances of Sync + Send trait implementation.

Be cautious with manual implementations of `Send` and `Sync` traits [[Rustbook: typesystem]](https://anssi-fr.github.io/rust-guide/06_typesystem.html#send-and-sync-traits) [[Rustbook: send and sync]](https://doc.rust-lang.org/nomicon/send-and-sync.html).
Both traits are _unsafe traits_, i.e., the Rust compiler does not verify in any way that they are implemented correctly. The danger is real: an incorrect implementation may lead to **undefined behavior**.

In the majority of scenarios, manual implementation is unnecessary. In Rust, nearly all primitive types intrinsically implement Send and Sync traits, and for a significant proportion of compound types, the Rust compiler automatically derives these implementations.

### Comparison Traits

> Ensure the implementation of standard comparison traits respects documented invariants [[comparison traits partialeq eq partialord ord]](https://anssi-fr.github.io/rust-guide/06_typesystem.html#comparison-traits-partialeq-eq-partialord-ord).

Reviewer 2: same question about relevancy - also what are the "documented invariants" and what does it mean within the aptos core context?

Author: Updated with a definition of invariants in Comparison Traits

> The ANSSI resource extensively covers the matter, an example in [Appendix](#implementation-of-comparison-traits) is provided as a really basic demonstration of implementation.

### Enums

Prefer enums for state management to prevent invalid state representation.

### Concurrency Safe Primitives

Make use of Rust’s concurrency primitives like `Arc`, `Mutex`, and `RwLock` to manage shared state [[Rustbook: concurrency]](https://doc.rust-lang.org/book/ch16-00-concurrency.html).
By utilizing these primitives, Rust programs can manage shared resources among multiple threads safely and efficiently, adhering to Rust's goals of enabling fearless concurrency and memory safety. This is critical in multithread programs because mistakes in concurrency can be very costly, both for the system's stability and security.

### Data Structures with Deterministic Internal Order

> Certain data structures, like HashMap and HashSet, do not guarantee a deterministic order for the elements stored within them. This lack of order can lead to problems in operations that require processing elements in a consistent sequence across multiple executions. Below is a list of deterministic data structures available in Rust. Please note, this list may not be exhaustive:

Reviewer 2: can we list some common scenarios in aptos when deterministic data structures are important?

Author: Added: In the Aptos blockchain, deterministic data structures help in achieving consensus, maintaining the integrity of the ledger, and ensuring that computations can be reliably reproduced across different nodes.

> - **BTreeMap:** It maintains its elements in sorted order by their keys.

Reviewer 2: ```suggestion
- **BTreeMap:** maintains its elements in sorted order by their keys.
```

- **BinaryHeap:** It maintains its elements in a heap order, which is a complete binary tree where each parent node is less than or equal to its child nodes.

> - **VEC**: It maintains its elements in the order in which they were inserted. ⚠️

Reviewer 2: ```suggestion
- **Vec**: It maintains its elements in the order in which they were inserted. ⚠️
```

- **LinkedList:** It maintains its elements in the order in which they were inserted. ⚠️
- **VecDeque:** It maintains its elements in the order in which they were inserted. ⚠️

## Cryptography

### Talk to a cryptographer

> If you’re using cryptography in your code and you are **not** a cryptographer, you don’t know what you don’t know (_e.g.,_ small subgroup attacks, malleability attacks, domain-separation attacks, collision resistance versus one-wayness, safe key reuse, correctly using randomness, nonce reuse, CCA security, CPA security, etc). So ask for help from a cryptographer in our team (Alin, Benny, Michael, Rex).

Reviewer 2: this guidance is not actionably for anyone that doesn't work at aptos labs. these guidelines are public and should be applicable to everonye. I would remove this.

Author: I think this came from someone else, I noticed the change while porting, but left while thinking for a better rephrasing.

### Zeroing Sensitive Data

Use [zeroize](https://docs.rs/zeroize/latest/zeroize/#) for zeroing memory containing sensitive data.

## Misc

### Forget and Memory Leaks

Avoid using `std::mem::forget` in secure development, or any other function that leaks the memory [[forget and memory leaks]](https://anssi-fr.github.io/rust-guide/05_memory.html#forget-and-memory-leaks).

> Reference cycles can also cause memory leakage [[Rustbook: leak]](https://doc.rust-lang.org/book/ch15-06-reference-cycles.html?highlight=leak#reference-cycles-can-leak-memory).

Most memory leaks result in general product reliability problems. If an attacker can intentionally trigger a memory leak, the attacker might be able to launch a denial-of-service attack (by crashing or hanging the program).

### Fuzzing
Aptos contains harnesses for fuzzing crash-prone code like deserializers, using [`libFuzzer`](https://llvm.org/docs/LibFuzzer.html) through [`cargo fuzz`](https://rust-fuzz.github.io/book/cargo-fuzz.html). For more examples, see the `testsuite/fuzzer` directory where find detailed README.md.

## Appendix

> #### Implementation of Comparison Traits

Reviewer 2: not sure how relevant this is

Author: We have 109+ trait implementation in our codebase, e.g. 27 PartialEq

Reviewer 2: let me rephrase - I think this sample code doesn't provide much value to the overall document, as we don't expect developers copying this when implementing comparison traits. additionally, there are no other code snippets related to other issues.
